src/stack.rs

```diff
@@ -1,14 +1,117 @@
-// Alex Ozdemir <aozdemir@hmc.edu> // <- Your name should replace this line!
+// Dan Obermiller <dobermiller16@cmc.edu> // <- Your name should replace this line!
 // Starter code for HMC's MemorySafe, week 1
 //
 // The implementation of SinglyLinkedList
 
+use std::mem;
 use Stack;
 
+struct Node<T> {
+    pointer: Option<Box<Node<T>>>,
+    data: T,
+}
+
 pub struct SinglyLinkedList<T> {
-    data: T // Probably not what you want
+    head: Option<Box<Node<T>>>,
+    size: usize,
 }
 
 impl<T: Eq> Stack<T> for SinglyLinkedList<T> {
-    // ????
+    fn new() -> Self {
+        SinglyLinkedList{head: None, size: 0}
+    }
+
+    fn push_front(&mut self, item: T) {
+        let old = mem::replace(&mut self.head, None);
+        self.head = Some(Box::new(Node { pointer: old, data: item }));
+        // let mut new_head = Node { pointer: None, data: item };
+        // mem::swap(&mut new_head.pointer, &mut self.head);
+        // self.head = Some(Box::new(new_head));
+        self.size += 1;
+    }
+
+    fn pop_front(&mut self) -> Option<T> {
+        self.head.take().map(|mut n| {
+            assert!(self.size > 0);
+            self.head = n.pointer.take();
+            self.size -= 1;
+            n.data
+        })
+    }
+
+    fn peek_front(&self) -> Option<&T> {
+        match self.head {
+            Some(ref node) => Some(&node.data),
+            None => None
+        }
+    }
+
+    fn len(&self) -> usize { self.size }
+
+    fn remove_first(&mut self, item: &T) -> Option<T> {
+        // I wanted this to work, but it didn't :(
```

Comment:
I believe that you can make this "traversal style" work, but you have to be very careful about who owns which node when (what the data structure and stack variables own during each step). Pictures are helpful for this.

Essentially you want to take ownership of rest of the list at the beginning and walk through it, adding one node at a time to the back of the list. Likely the only &mut you should maintain is to the back of the list.